core/src/main/scala/org/apache/spark/network/netty/NettyBlockTransferService.scala

```diff
@@ -48,6 +54,41 @@ class NettyBlockTransferService(conf: SparkConf, securityManager: SecurityManage
   private[this] var clientFactory: TransportClientFactory = _
   private[this] var appId: String = _
 
+  override def metricsCollection(executorId: String, executorMetrics: ExecutorMetrics): Unit = {
+    logDebug("starting a new thread for metricsCollection()")
+    new Thread("metric thread for executor " + executorId) {
+      override def run() {
+        while (true) {
+          getNettyMemMetrics(executorMetrics)
+          Thread.sleep(2000)
+        }
+      }
+    }.start()
+  }
+
+  private def getNettyMemMetrics(executorMetrics: ExecutorMetrics): Unit = {
+    val pooledAllocator = clientFactory.getPooledAllocator()
+    val directHeapSize: Long = sumOfMetrics(pooledAllocator.directArenas())
+    val onHeapSize: Long = sumOfMetrics(pooledAllocator.heapArenas())
+    executorMetrics.setNettyReadOnheapSize(onHeapSize)
+    executorMetrics.setNettyReadDirectheapSize(directHeapSize)
+    logInfo(s"current Netty directHeapSize is $directHeapSize, heapSize is $onHeapSize, " +
```

Comment:
probably needs to be quieter, maybe logDebug.  Though we just want to make the heartbeat have a logdebug for its metrics ...